### ISSUE TITLE
shim-15-ia32 for ROSA FRESH

### ISSUE THREAD

**Issue by keleg** [rhboot/shim-review#64]
Make sure you have provided the following information:

 - [x] link to your code branch cloned from rhboot/shim-review in the form user/repo@tag
  keleg/shim-review@ntcitrosa-shim-ia32-20190422
 - [x] completed README.md file with the necessary information
    https://github.com/keleg/shim-review/blob/ntcitrosa32/README.md
 - [x] shim.efi to be signed
  https://github.com/keleg/shim-review/blob/ntcitrosa-shim-ia32-20190422/shimia32.efi
 - [x] public portion of your certificate embedded in shim (the file passed to VENDOR_CERT_FILE)
  https://github.com/keleg/shim-review/blob/ntcitrosa-shim-ia32-20190422/rosa.cer
 - [x] any extra patches to shim via your own git tree or as files
  https://abf.io/signer/shim-unsigned
 - [x] any extra patches to grub via your own git tree or as files
  https://abf.io/import/grub2
 - [x] build logs
  https://abf.io/build_lists/3006835
  on github: https://github.com/keleg/shim-review/blob/ntcitrosa-shim-ia32-20190422/build-shim-ia32-3006835.log

###### What organization or people are asking to have this signed:
`LLC "NTC IT ROSA"`

###### What product or service is this for:
`"ROSA Fresh" - Linux Desktop`

###### What is the origin and full version number of your shim?
`https://github.com/rhboot/shim`

`We use v15 will all patches from that git repo up to rev. 20e731f4 inclusive.`

###### What's the justification that this really does need to be signed for the whole world to be able to boot it:
`ROSA Fresh is a non-profit Linux distribution developed by the community and has a long history. Is deployed on a high number of nodes already using it in SecureBoot mode enabled.`

###### How do you manage and protect the keys used in your SHIM?
`Shim has the public key of the EV Code Signing key pair (issued by DigiCert) built-in. The key is used to validate GRUB boot loader. No private keys are embedded.Shim binary itself is signed, so the built-in public key cannot be modified or removed without making the signature invalid. This guarantees that if shim has been tampered with and is then used in SecureBoot environments, this will be detected immediately.`

###### Do you use EV certificates as embedded certificates in the SHIM?
`Shim has the public key of the EV Code Signing key pair (issued by DigiCert) built-in.`

###### What is the origin and full version number of your bootloader (GRUB or other)?
`The source code of GRUB is available here: ftp://ftp.gnu.org/gnu/grub/grub-2.02.tar.xz <ftp://ftp.gnu.org/gnu/grub/grub-2.02.tar.xz>. The patches to GRUB 2.02 specific to ROSA Linux, as well as build scripts, are available here: https://abf.io/import/grub2 `

###### If your SHIM launches any other components, please provide further details on what is launched
`Apart from the boot loader (GRUB), shim can launch MokManager tool (developed alongside shim, https://github.com/rhboot/shim )`

###### How do the launched components prevent execution of unauthenticated code?
`Same as GRUB, MokManager executable must be signed with the appropriate key (EV Code Signing) for shim to validate and launch it. MokManager itself executes no unauthenticated code.`

###### Does your SHIM load any loaders that support loading unsigned kernels (e.g. GRUB)?
`Shim launches GRUB`

###### What kernel are you using? Which patches does it includes to enforce Secure Boot?
`Our current kernel is based on the kernel 4.15.0-47.50-generic from Ubuntu 18.04 LTS (http://kernel.ubuntu.com/git/kernel-ppa/mirror/ubuntu-bionic.git/), which already contains the patches to enforce SecureBoot as needed. We have no additional SecureBoot-related patches on top of that.`

`Our patches, configs and build instructions for the kernel (RPM spec file) are available here: https://abf.io/import/kernel-desktop-4.15 `

###### What changes were made since your SHIM was last signed?
```
This is an update from v0.9 to v15 + changes up to rev. 20e731f4 from the upstream git repo.
```

###### What is the hash of your final SHIM binary?
`sha256: 91108c0b59386783d58c0f1ef9351c21e6fc2ebaaf3e8af5f62b8c312f98d331  shimia32.efi`

**Comment by steve-mcintyre** [comment]
Hi!

Looking through your review requests now. Apologies for the time taken - I'm new to the team reviewing things here. A few questions...

1. Using your VM to build, I'm repeatably seeing the following checksum: 2b5923703f25d3f2e11da0fe947a1ae11215ea195719bd0909a3400d21cdd8ca  shimia32.efi

but your submitted binary is
91108c0b59386783d58c0f1ef9351c21e6fc2ebaaf3e8af5f62b8c312f98d331  shimia32.efi

Can you explain the difference please?

2. What patches do you have applied to your version of Grub? The link to https://abf.io/import/grub2 doesn't work for me. :-(

3. The review template asks for PGP keys for each of the contacts listed. Do you have those please?

**Comment by euspectre** [comment]
Hi,

> 1. Using your VM to build, I'm repeatably seeing the following checksum: 2b5923703f25d3f2e11da0fe947a1ae11215ea195719bd0909a3400d21cdd8ca shimia32.efi

I see. '.git' directory is still present in the source tree, 'make'  detects that and embeds the git revision in the binary. I have compared the binaries - yes, the submitted shim binary has the string "master" embedded while the rebuilt binary has the git revision there: "20e731f423a438f53738de73af9ef3d67c4cba2f". The remaining differences appeared because the size of the section containing the string differs and the offsets of the subsequent sections differ as a result.

You can remove .git directory and try the build again. I've got the same shim binary this way now.
This also applies to https://github.com/rhboot/shim-review/issues/63.

Not sure if there are any timestamps in the code which might also differ. If that happens, it should be not difficult to check.

> 2. What patches do you have applied to your version of Grub? The link to https://abf.io/import/grub2 doesn't work for me. :-(

Might be some tech difficulties at abf.io. I cloned the repo here, for convenience: https://github.com/euspectre/grub2-rosa-new/tree/for-rosa-r12

The list of patches and the order they are applied in can be seen in the RPM spec file: https://github.com/euspectre/grub2-rosa-new/blob/for-rosa-r12/grub2.spec#L30

Basically, ROSA uses
* some  patches from upstream GRUB, 
* a few dozen patches from RHEL, Fedora and Ubuntu (this group includes the patches for proper SecureBoot support),
* patches from Virtuozzo/VZLinux (mostly - theme-related)
* some ROSA-specific patches.

From the last group, only one patch is related to SB: https://github.com/euspectre/grub2-rosa-new/blob/for-rosa-r12/grub2-Install-signed-images-if-UEFI-Secure-Boot-is-enabled.patch. This one was originally taken from Ubuntu but reworked for ROSA.

**Comment by euspectre** [comment]
> 3. The review template asks for PGP keys for each of the contacts listed. Do you have those please?

@keleg : Vladimir, you did not provide the public keys this time. Please ask each security contact you listed to provide their public keys, place the files into the appropriate branches of your git repo with shim and list the links to the files in README.md.

**Comment by steve-mcintyre** [comment]
> Hi,
> 
> > 1. Using your VM to build, I'm repeatably seeing the following checksum: 2b5923703f25d3f2e11da0fe947a1ae11215ea195719bd0909a3400d21cdd8ca shimia32.efi
> 
> I see. '.git' directory is still present in the source tree, 'make' detects that and embeds the git revision in the binary. I have compared the binaries - yes, the submitted shim binary has the string "master" embedded while the rebuilt binary has the git revision there: "20e731f423a438f53738de73af9ef3d67c4cba2f". The remaining differences appeared because the size of the section containing the string differs and the offsets of the subsequent sections differ as a result.
> 
> You can remove .git directory and try the build again. I've got the same shim binary this way now.
> This also applies to #63.

ACK, I get the same now too. Good call. :-)

> > 1. What patches do you have applied to your version of Grub? The link to https://abf.io/import/grub2 doesn't work for me. :-(
> 
> Might be some tech difficulties at abf.io. I cloned the repo here, for convenience: https://github.com/euspectre/grub2-rosa-new/tree/for-rosa-r12

OK, cool. Thanks!

> The list of patches and the order they are applied in can be seen in the RPM spec file: https://github.com/euspectre/grub2-rosa-new/blob/for-rosa-r12/grub2.spec#L30
> 
> Basically, ROSA uses
> 
>     * some  patches from upstream GRUB,
> 
>     * a few dozen patches from RHEL, Fedora and Ubuntu (this group includes the patches for proper SecureBoot support),
> 
>     * patches from Virtuozzo/VZLinux (mostly - theme-related)
> 
>     * some ROSA-specific patches.
> 
> 
> From the last group, only one patch is related to SB: https://github.com/euspectre/grub2-rosa-new/blob/for-rosa-r12/grub2-Install-signed-images-if-UEFI-Secure-Boot-is-enabled.patch. This one was originally taken from Ubuntu but reworked for ROSA.

Nod. What's your policy when Secure Boot is enabled - will your grub allow fall-through to an unsigned kernel image like Ubuntu have done?

**Comment by euspectre** [comment]
> Nod. What's your policy when Secure Boot is enabled - will your grub allow fall-through to an unsigned kernel image like Ubuntu have done?

GRUB currently used in ROSA R11 allows such fall-through.

The future versions of GRUB (to be used in ROSA R12 or newer with shim discussed here) are likely to disable it and disallow unsigned kernels when SB is active. The sources of GRUB referred to earlier were prepared with that in mind.

It is up to ROSA tech committee to decide though, but ROSA follows Ubuntu here, to some extent. As it was pointed out in another review request, Ubuntu no longer allows fallback to unsigned kernels. So, I expect the committee to propose using only signed kernels for the flavours of ROSA with enhanced security and recommend the users of other ROSA flavours to disable SB.

**Comment by keleg** [comment]
>     2\. The review template asks for PGP keys for each of the contacts listed. Do you have those please?
I add PGP fingerprints to Readme.md for ia32 and x64. It's enough?

**Comment by steve-mcintyre** [comment]
Discussing this with other shim reviewers, we **really** want to see distributors disabling fallback to unsigned kernels. The Ubuntu developers have stopped doing this some time ago, and I'm not aware of any other distros who support this any more.

**Comment by euspectre** [comment]
> we really want to see distributors disabling fallback to unsigned kernels.

As I wrote before, the updated GRUB, which is to be used with this shim in ROSA R12, does disable the fallback. What else?

As for the older versions of ROSA, including R11, they have been released with the older shim and are of no concern here.

**Comment by euspectre** [comment]
> The Ubuntu developers have stopped doing this some time ago

If I understand the changelogs from Ubuntu's GRUB packages correctly, they did it in v2.02-2ubuntu8.11 for Bionic, 05 Feb 2019.

**Comment by steve-mcintyre** [comment]
ACK, OK.

**Comment by steve-mcintyre** [comment]
Based on this conversation, I think we're good to go. Apologies it's taken so long, as I said earlier!

**Comment by keleg** [comment]
Thank you! 
And https://github.com/rhboot/shim-review/issues/63?
We support 32-uefi and 64 uefi loaders.

**Comment by steve-mcintyre** [comment]
Nod, just about to mark that one as well.
